File: langs/typescript/src/collections/servicedObservableCollection.ts
/**
 * ServicedObservableCollection<T> — observable array that optionally publishes
 * CollectionChangedMessage events to an IMessageHub.
 *
 * When no hub is injected, the class behaves like a plain observable collection
 * (local RxJS Subject emits events; no publication, no errors).
 *
 * Ownership stays with the caller: removing, replacing, or clearing an item does
 * not call dispose/destruct or any VM lifecycle method on that item.
 *
 * See spec/21-collections.md §2, ADR-0024, and ADR-0096.
 */
import { Observable, Subject } from "rxjs";
import type { IMessageHub } from "../services/messageHub.js";
import { CollectionChangedMessage } from "../messages/collectionChanged.js";

export { CollectionChangedMessage };

export class ServicedObservableCollection<T> {
  readonly #hub: IMessageHub | null;
  #items: T[] = [];
  readonly #subject = new Subject<CollectionChangedMessage<T>>();

  constructor(hub?: IMessageHub | null) {
    this.#hub = hub ?? null;
  }

  // ── Public surface ──────────────────────────────────────────────────────

  /** Hot observable of CollectionChangedMessage events. */
  get collectionChanged(): Observable<CollectionChangedMessage<T>> {
    return this.#subject.asObservable();
  }

  get length(): number {
    return this.#items.length;
  }

  /** Return the item at *index*. */
  at(index: number): T | undefined {
    return this.#items[index];
  }

  /** Return a shallow copy of the backing array. */
  toArray(): T[] {
    return [...this.#items];
  }

  [Symbol.iterator](): IterableIterator<T> {
    return this.#items[Symbol.iterator]();
  }

  // ── Mutations ───────────────────────────────────────────────────────────

  /** Append *item* to the end. */
  push(item: T): void {
    const index = this.#items.length;
    this.#items.push(item);
    this.#emit(CollectionChangedMessage.forAdd(this, item, index));
  }

  /** Remove and return the last item, or undefined if empty. */
  pop(): T | undefined {
    if (this.#items.length === 0) return undefined;
    const index = this.#items.length - 1;
    const item = this.#items[index] as T;
    this.#items.pop();
    this.#emit(CollectionChangedMessage.forRemove(this, item, index));
    return item;
  }

  /** Remove the first indexOf match for *item*. */
  remove(item: T): boolean {
    const index = this.#items.indexOf(item);
    if (index === -1) return false;
    this.removeAt(index);
    return true;
  }

  /** Remove the item at *index*. */
  removeAt(index: number): void {
    this.#validateIndex(index);
    const item = this.#items[index] as T;
    this.#items.splice(index, 1);
    this.#emit(CollectionChangedMessage.forRemove(this, item, index));
  }

  /**
   * Remove *count* items starting at *start* and optionally insert *newItems*.
   * Emits a Reset message (coarse-grained) for any multi-item splice.
   * For single-item operations (no inserts, count=1) emits a Remove.
   */
  splice(start: number, deleteCount?: number, ...newItems: T[]): T[] {
    const lengthBeforeSplice = this.#items.length;
    const removed = this.#items.splice(start, deleteCount ?? this.#items.length, ...newItems);
    if (removed.length === 0 && newItems.length === 0) {
      // No-op splice: nothing mutated, so nothing is emitted
      // (spec/21 §2.4 — messages are emitted per mutation).
      return removed;
    }
    if (removed.length === 1 && newItems.length === 0) {
      // Normalize the start index the way Array.prototype.splice resolves it
      // (negatives count from the end, out-of-range clamps) so the emitted
      // Remove carries the actual removal position (spec/21 line 148,
      // "indexBeforeRemoval"), not a raw negative/out-of-range argument.
      const resolvedStart =
        start < 0
          ? Math.max(lengthBeforeSplice + start, 0)
          : Math.min(start, lengthBeforeSplice);
      this.#emit(
        CollectionChangedMessage.forRemove(this, removed[0] as T, resolvedStart),
      );
    } else {
      this.#emit(CollectionChangedMessage.forReset(this));
    }
    return removed;
  }

  /** Replace the item at *index*, emitting even when the value is identical. */
  replace(index: number, newItem: T): void {
    this.#validateIndex(index);
    const oldItem = this.#items[index] as T;
    this.#items[index] = newItem;
    this.#emit(
      CollectionChangedMessage.forReplace(this, newItem, oldItem, index),
    );
  }

  /** Source-compatible alias for replace. */
  setAt(index: number, newItem: T): void {
    this.replace(index, newItem);
  }

  /** Replace the complete contents from a snapshot of *items*. */
  replaceAll(items: Iterable<T>): void {
    const snapshot = [...items];
    if (this.#items.length === 0 && snapshot.length === 0) return;
    this.#items = snapshot;
    this.#emit(CollectionChangedMessage.forReset(this));
  }

  /** Move an item using pre-move source and destination indices. */
  move(fromIndex: number, toIndex: number): void {
    this.#validateIndex(fromIndex);
    this.#validateIndex(toIndex);
    if (fromIndex === toIndex) return;

    const item = this.#items[fromIndex] as T;
    this.#items.splice(fromIndex, 1);
    this.#items.splice(toIndex, 0, item);
    this.#emit(
      CollectionChangedMessage.forMove(this, item, fromIndex, toIndex),
    );
  }

  /** Remove all items. */
  clear(): void {
    if (this.#items.length === 0) return;
    this.#items.length = 0;
    this.#emit(CollectionChangedMessage.forReset(this));
  }

  // ── Internal ────────────────────────────────────────────────────────────

  #emit(msg: CollectionChangedMessage<T>): void {
    // 1. Notify local subscribers.
    this.#subject.next(msg);
    // 2. Publish to hub (when present).
    this.#hub?.send(msg);
  }

  #validateIndex(index: number): void {
    if (
      !Number.isInteger(index) ||
      index < 0 ||
      index >= this.#items.length
    ) {
      throw new RangeError(`Index ${String(index)} out of bounds`);
    }
  }
}

File: langs/swift/Sources/VMx/Collections/ServicedObservableCollection.swift
//
// ServicedObservableCollection — observable array that optionally publishes
// CollectionChangedMessage events to a MessageHubProtocol.
//
// When no hub is injected (hub == nil), the class behaves like a plain
// observable collection: the local collectionChanged publisher fires but
// no hub publication occurs and no error is raised (COL-003).
//
// Ordering guarantee (COL-001/002): local collectionChanged fires FIRST,
// then hub?.send(_:). Delivery is synchronous on the caller's thread,
// with no dispatcher marshaling (COL-004).
//
// Ownership stays with the caller: removing, replacing, or clearing an item
// does not call dispose/destruct or any VM lifecycle method on that item.
//
// See spec/21-collections.md §2, ADR-0024, and ADR-0096.
//
import Combine
import Foundation

/// Hub-aware observable collection. Each mutation:
/// 1. Emits a `CollectionChangedMessage` on the local `collectionChanged` publisher.
/// 2. Sends the same message to the injected `MessageHubProtocol` (if any).
///
/// Both steps happen synchronously on the caller thread (COL-004).
public final class ServicedObservableCollection<T>: Sequence {

    // ── Private state ────────────────────────────────────────────────────────

    private let hub: (any MessageHubProtocol)?
    private var items: [T] = []
    private let subject = PassthroughSubject<CollectionChangedMessage<T>, Never>()
    private let typeName: String

    // ── Initializer ──────────────────────────────────────────────────────────

    /// Create a serviced collection optionally backed by a hub.
    ///
    /// - Parameter hub: The hub to publish `CollectionChangedMessage` events to.
    ///   Pass `nil` (the default) for a hub-less observable collection.
    public init(hub: (any MessageHubProtocol)? = nil) {
        self.hub = hub
        self.typeName = "ServicedObservableCollection"
    }

    // ── Public surface ───────────────────────────────────────────────────────

    /// Hot publisher of `CollectionChangedMessage` events for this collection.
    /// Subscribers see only events emitted after they subscribe.
    public var collectionChanged: AnyPublisher<CollectionChangedMessage<T>, Never> {
        subject.eraseToAnyPublisher()
    }

    /// The number of items in the collection.
    public var count: Int { items.count }

    /// Return the item at `index`, or `nil` if out of bounds.
    public func at(_ index: Int) -> T? {
        guard index >= 0 && index < items.count else { return nil }
        return items[index]
    }

    /// Return a shallow copy of the backing array.
    public func toArray() -> [T] { items }

    /// Iterate over a snapshot of the current contents.
    public func makeIterator() -> IndexingIterator<[T]> {
        items.makeIterator()
    }

    // ── Mutations ────────────────────────────────────────────────────────────

    /// Append `item` to the end of the collection.
    public func append(_ item: T) {
        let index = items.count
        items.append(item)
        emit(.forAdd(sender: self, senderName: typeName, item: item, index: index))
    }

    /// Remove and return the last item, or `nil` if the collection is empty.
    @discardableResult
    public func removeLast() -> T? {
        guard !items.isEmpty else { return nil }
        let index = items.count - 1
        let item = items[index]
        items.removeLast()
        emit(.forRemove(sender: self, senderName: typeName, item: item, index: index))
        return item
    }

    /// Remove the item at `index`.
    ///
    /// - Precondition: `index` is a valid index in the collection.
    public func removeAt(_ index: Int) {
        precondition(index >= 0 && index < items.count,
                     "ServicedObservableCollection.removeAt: index \(index) out of bounds (count: \(items.count))")
        let item = items.remove(at: index)
        emit(.forRemove(sender: self, senderName: typeName, item: item, index: index))
    }

    /// Replace the item at `index` with `newItem`.
    ///
    /// - Precondition: `index` is a valid index in the collection.
    public func replace(at index: Int, with newItem: T) {
        precondition(index >= 0 && index < items.count,
                     "ServicedObservableCollection.replace: index \(index) out of bounds (count: \(items.count))")
        let oldItem = items[index]
        items[index] = newItem
        emit(.forReplace(sender: self, senderName: typeName,
                         newItem: newItem, oldItem: oldItem, index: index))
    }

    /// Replace the item at `index` with `newItem`.
    ///
    /// - Precondition: `index` is a valid index in the collection.
    public func setAt(_ index: Int, _ newItem: T) {
        replace(at: index, with: newItem)
    }

    /// Replace all contents from a fully materialized input snapshot.
    public func replaceAll<S: Sequence>(_ newItems: S) where S.Element == T {
        let snapshot = Array(newItems)
        guard !items.isEmpty || !snapshot.isEmpty else { return }
        items = snapshot
        emit(.forReset(sender: self, senderName: typeName))
    }

    /// Move an existing item between two pre-move positions.
    public func move(from oldIndex: Int, to newIndex: Int) throws {
        guard oldIndex >= 0 && oldIndex < items.count else {
            throw VMCollectionIndexError(index: oldIndex, count: items.count)
        }
        guard newIndex >= 0 && newIndex < items.count else {
            throw VMCollectionIndexError(index: newIndex, count: items.count)
        }
        guard oldIndex != newIndex else { return }

        let item = items.remove(at: oldIndex)
        items.insert(item, at: newIndex)
        emit(.forMove(sender: self, senderName: typeName, item: item,
                      from: oldIndex, to: newIndex))
    }

    /// Remove all items and emit a Reset event.
    public func clear() {
        guard !items.isEmpty else { return }
        items.removeAll()
        emit(.forReset(sender: self, senderName: typeName))
    }

    // ── Internal ─────────────────────────────────────────────────────────────

    /// Emit locally first, then publish to the hub (COL-001/002 ordering).
    private func emit(_ message: CollectionChangedMessage<T>) {
        // 1. Local subscribers first.
        subject.send(message)
        // 2. Hub publication (no-op when hub is nil — COL-003).
        hub?.send(message)
    }
}

// MARK: - Value-based removal (requires Equatable)

extension ServicedObservableCollection where T: Equatable {
    /// Remove the first occurrence of `item`, emitting a granular remove (local
    /// then hub, COL-001/002 ordering) exactly like the other mutators. Returns
    /// `true` when the item was found and removed, `false` otherwise — parity with
    /// the canonical `Remove(item): Bool` (spec/21 §2.1). Offered on an
    /// `Equatable`-constrained extension because the element type is unconstrained.
    @discardableResult
    public func remove(_ item: T) -> Bool {
        guard let index = items.firstIndex(of: item) else { return false }
        let storedItem = items.remove(at: index)
        emit(.forRemove(sender: self, senderName: typeName, item: storedItem, index: index))
        return true
    }
}

// MARK: - Aggregate membership source (reference items only)

extension ServicedObservableCollection: ObservableMembershipSource where T: AnyObject {
    public typealias Item = T

    public func snapshot() -> [T] { toArray() }

    public func subscribeMembership(_ callback: @escaping () -> Void) -> AnyCancellable {
        collectionChanged.sink { _ in callback() }
    }
}
